Guide for Stellar smart contract init templates (#806)
stellar/stellar-docs: docs/build/guides/conventions/soroban-contract-init-template.mdx

```bash
npm create vite@latest soroban-react-template --template react
cd soroban-react-template
npm install
npm run dev
```

## Creating the initialize script

The initialization script is the heart of the contract template. This little but highly crucial script enables us to execute Stellar CLI commands which, in turn, enables us to build, deploy, and generate smart contract bindings.

Usually, the initialization script is written in bash script but JavaScript can be used as well. This guide uses the latter but the same principles are involved when using bash scripts.

We start off by creating `initialize.js` files in our frontend project directory.

```bash
touch initialize.js
```

This code creates the `initailze.js` file. Now let's open the file to begin writing the necessary code.

### Installing the dependencies

The following dependencies will simplify the process of interacting with files and directories which the `initialize.js` file does together with other functionalities.

1. [glob](https://npmjs.com/package/glob)
2. [dotenv](https:npmjs.com/package/dotenv)

Install the above packages as dev dependencies by running the code below:

> ```bash

Reviewer: ````suggestion
```bash npm2yarn
````

Just a suggestion, not a "must-have." You can add this to your `bash` code blocks when you are using `npm` or `yarn` commands and it'll provide the options for both, depending on what the user selects.

Reviewer: I don't know for sure if it would work in the section above where you're create the project, installing, etc.

npm install dotenv glob --dev
```

### Writing the initialization code

As a general overview, the following script will perform the below tasks:

1. Create network profiles for contract project
2. Generate and fund account for interaction with the Stellar network
3. Build smart contract .wasm files
4. Deploy smart contract to the Stellar network
5. Generate smart contract Typescript bindings

#### Importing the modules

The following modules allow us to work with environment variables, execute Stellar CLI commands, search files and directories, work with paths, and create and delete files, in that order.

```javascript
import "dotenv/config";
import { execSync } from "child_process";
import { globSync } from "glob";
import path from "path";
import { writeFileSync, rmSync } from "fs";
```

### Creating the wrapper function

Instead of calling the `execSync` function and making the error handling process multiple times, we create a wrapper function to encapsulate this function.

> Furthermore, when deploying the contract, the command returns the deployed contract ID, which is required when generating contract Typescript bindings. Wrapping the `execSync` function will enable to return Stellar CLI command output while handling errors in one go.

Reviewer: ```suggestion
Furthermore, when deploying the contract, the command returns the deployed contract address, which is required when generating contract Typescript bindings. Wrapping the `execSync` function will enable the script to return Stellar CLI command output while handling errors in one go.
```

    console.error("Error executing command: \n", error.message);
    return null; // or throw error if you want to propagate it
  }
};
```

### Creating network profiles

This code snippet executes a Stellar CLI command to add a network profile. It uses environment variables (process.env.NETWORK, process.env.RPC_URL, process.env.NETWORK_PASSPHRASE) to specify the network details and RPC URL.

You can clearly see the benefit of wrapping the `execSync` function as it results in a much cleaner code and short snippet while retaining same functionality.

```javascript
run(
  `stellar network add ${process.env.NETWORK} --rpc-url ${process.env.RPC_URL} --network-passphrase ${process.env.NETWORK_PASSPHRASE}`,
);
```

### Creating account profiles

This code generates Stellar key pairs for an account (process.env.ACCOUNT) and funds them accordingly on a specific network (process.env.NETWORK).

```javascript
run(
  `stellar keys generate ${process.env.ACCOUNT} --network ${process.env.NETWORK}`,
);
```

### Building the contract

This snippet builds Stellar smart contracts in the project directory.

```javascript
run("stellar contract build");
```

### Deploying the contract

> This code deploys one or more Stellar smart contracts. It uses files found in target/wasm32-unknown-unknown/release/\*.wasm, deploys each with specified parameters, and stores contract details (alias and contractId) in contractsObject.

Reviewer: ```suggestion
This code deploys one or more Stellar smart contracts. It uses files found in `target/wasm32-unknown-unknown/release/*.wasm`, deploys each with specified parameters, and stores contract details (alias and contractId) in contractsObject.
```

When writing filepaths or filenames (outside of scripts), let's use backticks to separate them and make it a bit easier for the eye to catch what's happening.

The code iterates through the smart contract files found in the project directory and deploys all of them.

```javascript
const files = globSync("target/wasm32-unknown-unknown/release/*.wasm");
let alias, contractId;
console.log(files);
files.forEach((file) => {
  alias = path.basename(file).split(".")[0];
  console.log(`Deploying Contract ${alias}`);
  contractId = run(
    `stellar contract deploy --network ${process.env.NETWORK} --source ${process.env.ACCOUNT} --wasm ${file} --alias ${alias}`,
  );
  //save contractId and alias to contracts Object
  contractsObject[alias] = {
    alias,
    id: contractId,
  };
});
```

### Generating TypeScript bindings

> This snippet generates TypeScript bindings for Stellar smart contracts deployed earlier. It iterates over files and uses `contractsObject` to obtain contract details (alias and `contractId`).

Reviewer: ```suggestion
This snippet generates TypeScript bindings for Stellar smart contracts deployed earlier. It iterates over files and uses `contractsObject` to obtain contract details (`alias` and `contractId`).
```

```javascript
let filename, output_dir;
files.forEach((file) => {
  filename = path.basename(file).split(".")[0];
  output_dir = `./packages/${filename}`;
  run(
    `stellar contract bindings typescript --output-dir ${output_dir} --network ${process.env.NETWORK} --contract-id ${contractsObject[alias]["id"]}  --wasm ${file} --overwrite`,
  );
});
```

> ### Complete initialize.js script

Reviewer: ```suggestion
### Complete initialization script
```

> The following snippet is the complete aggregated version of the snippets shown earlier. I've added some functionalities like deleting pre-existing contract .wasm files and creating contract.json file, which are all optional steps.

Reviewer: ```suggestion
The following snippet is the complete aggregated version of the snippets shown earlier. I've added some functionalities like deleting pre-existing contract Wasm files and creating a `contract.json` file, which are all optional steps.
```

//Deploy Contract Wasm and save the contact Ids and Aliases
const files = globSync("target/wasm32-unknown-unknown/release/*.wasm");
let alias, contractId;
console.log(files);
files.forEach((file) => {
  alias = path.basename(file).split(".")[0];
  console.log(`Deploying Contract ${alias}`);
  contractId = run(
    `stellar contract deploy --network ${process.env.NETWORK} --source ${process.env.ACCOUNT} --wasm ${file} --alias ${alias}`,
  );
  //save contractId and alias to contracts Object
  contractsObject[alias] = {
    alias,
    id: contractId,
  };
});

//Create a json File to store the contract name as key and contract id and contract alias contained in an object as value
const contractsJSONPath = "./contracts.json";
const contractsJSONData = JSON.stringify(contractsObject);

try {
  writeFileSync(contractsJSONPath, contractsJSONData);
  console.log(`JSON file has been saved to ${contractsJSONPath}`);
} catch (err) {
  console.error("Error writing JSON file: ", err);
}

//Generate Contract Typescript Bindings
let filename, output_dir;
files.forEach((file) => {
  filename = path.basename(file).split(".")[0];
  output_dir = `./packages/${filename}`;
  run(

>     `stellar contract bindings typescript --output-dir ${output_dir} --network ${process.env.NETWORK} --contract-id ${contractsObject[alias]["id"]}  --wasm ${file} --overwrite`,

Reviewer: Is there a reason we're using double-quotes around `["id"]`, but not around `[alias]` when we're referencing the `--contract-id`? Unless there's a reason to prevent it, let's do the same in both, whichever way you want.

Author: alias in this case is a variable whose value we want to use while id is the key we are referencing. putting alias in quotes will break the app

  );
});
```

## Modifying `package.json` file

After saving the `initialize.js` script, let's modify our `package.json` file to create a new entry under scripts to run the code to work on the contract files.

Let's also modify the file to create a new workspace in the `package.json` to point to the path where the the TypeScript bindings are generated

```json
"scripts": {
    "setup": "node initialize.js",
  },

 "workspace": ["./packages/*"],
```

## Running the script

Now that we have a complete script and a frontend, we should try to run the script. However, because the script is dependent on contract files we need to have a contract project ready, otherwise the script will fail and result in an error. To solve this issue we have to either copy the whole project contents and paste them to an existing Stellar contract project directory or upload the template file to a repo then create a new project and setting the 'f' parameter to point to the hosted repository of the template.

### Setting up the project

> I have hosted a [public repository of the React template](https://github.com/CaptainPrinz/soroban-react-template). The following snippet implements the second option:

Reviewer: ```suggestion
A community member has hosted a [public repository of the React template](https://github.com/CaptainPrinz/soroban-react-template) available for use. The following command implements the second option:
```

```bash
stellar contract init token_contract -w token -f https://github.com/CaptainPrinz/soroban-react-template.git
```

> The snippet above initializes a new Soroban project, `token_contract` that uses a prebuild token contract and adds the frontend template we deployed to a remote repository.

Reviewer: ```suggestion
The snippet above initializes a new Stellar smart contracts project, `token_contract`, that uses a prebuilt token contract and adds the frontend template we deployed to a remote repository.
```

### Setting the environment variables

> Our templates offer a great deal of customization via environment variables. These variables need to be set by the user by creating a `.env` file otherwise the process will fail as these variables will be undefined.

Reviewer: Let's add a `.env.example` file in the repository, and discuss here how a dev can copy/move that file to `.env` and customize it to fit their needs. You wouldn't need to go into great detail as to the "how," people will figure that out lol. You could also add a link to the example file here somewhere, if you wanted to.

### Starting the script

> On your terminal run the following code:

Reviewer: ```suggestion
On your terminal run the following command:
```

```bash
npm run setup
```

This runs the `initialize.js` script and first sets the network profile, account identities, builds the contract `.wasm` file, deploys the contracts, and generates the bindings.

If successful, there should be the following changes in the directory:

> 1. A new `.soroban` directory, which contains the network profile files, account identity files, and hosted smart contract files in json format

Reviewer: let's briefly point out how this directory should be added to `.gitignore`

Reviewer: Let's **also** add it to your repository's `.gitignore` file, to make sure it's setup for this by default

> 2. A `package` directory containing the generated TypeScript bindings

Reviewer: ```suggestion
2. A `packages` directory containing the generated TypeScript bindings
```

3. A `contracts.json` file

You are now finally able to import these generated bindings in your frontend application, create their clients, and use them to interact with the contracts deployed on the network.

## Customization

> With the `initialize.js` script being the heart of the template file, you can go through the file and modify the paths in which files are saved in if necessary.

Reviewer: ```suggestion
With the `initialize.js` script being the heart of the template file, you can go through the file and modify the paths in which files are saved if necessary.
```
Redundant `in`.

Also a great deal of customization is enabled by using environment variables, giving flexibility to change network profiles and account identity without limits.

## Conclusion

> This section concludes the guide. Check out the documentation on how bindings can be used in dappa as this is a template tutorial is limited to areas concerned with templates as such the processes of connecting these generated bindings and using them in frontend appications is not covered.

Reviewer: ```suggestion
This section concludes the guide. Check out the documentation on how bindings can be used in dapps as this is a template tutorial is limited to areas concerned with templates as such the processes of connecting these generated bindings and using them in frontend applications is not covered.
```
Couple spelling errors

Reviewer: This sentence is also a bit of a run-on. Can you clean up or rephrase this to be a little clearer?

After saving the `initialize.js` script, let's modify our `package.json` file to create a new entry under scripts to run the code to work on the contract files.

Let's also modify the file to create a new workspace in the `package.json` to point to the path where the the TypeScript bindings are generated

```json
"scripts": {
    "setup": "node initialize.js",
  },

 "workspace": ["./packages/*"],
```

## Running the script

Now that we have a complete script and a frontend, we should try to run the script. However, because the script is dependent on contract files we need to have a contract project ready, otherwise the script will fail and result in an error. To solve this issue we have to either copy the whole project contents and paste them to an existing Stellar contract project directory or upload the template file to a repo then create a new project and setting the 'f' parameter to point to the hosted repository of the template.

### Setting up the project

A community member has hosted a [public repository of the React template](https://github.com/CaptainPrinz/soroban-react-template) available for use. The following command implements the second option:

```bash

> stellar contract init `hello_contract` -w hello -f https://github.com/CaptainPrinz/soroban-react-template.git

Reviewer: ```suggestion
stellar contract init hello_contract -f https://github.com/CaptainPrinz/soroban-react-template.git
```
The stellar cli (as of `v21.2.0` this morning) will fail if you run this command as-is. The project directory (hello_contract) needs to _not_ be in back-ticks, and `-w hello` is unnecessary, as the hello world contract is included by default. The cli will give an error about it being specified, too.

Author: fixed

```

The snippet above initializes a new Stellar smart contracts project, `hello_contract`, that uses a prebuilt `hello_world` contract and adds the frontend template we deployed to a remote repository.

### Setting the environment variables

Our templates offer a great deal of customization via environment variables. These variables need to be set by the user by creating a `.env` file otherwise the process will fail as these variables will be undefined.

### Starting the script

On your terminal run the following code:

```bash

> npm run setup

Reviewer: Your template repository has a [git conflict saved](https://github.com/CaptainPrinz/soroban-react-template/blob/c2dbaf770005411c87aa2c9b1110aa202cd734fd/initialize.js#L30-L34) in the `initialize.js` file, so this step fails with an error.

Reviewer: Running through the process again, two things I've come across:

```suggestion
npm install
cp .env.example .env # edit .env file as needed
npm run setup
```

First, let's be **overt** and say `npm install` needs to run. It's a bit "goes without saying," but I think it'll be common enough for someone going through the tutorial step-by-step to forget about, so let's just be extra clear.

Second, if the `.env.example` file isn't copied over, the initialization script will error out and complain about network and account configs missing.

Third, I'm getting a couple errors when running `npm run setup`. Here's all the output:

```bash
❯ npm run setup          

> hello_contract@0.0.0 setup
> node initialize.js

Configuring Network
Error executing command: 
 Command failed: stellar network add testnet --rpc-url http://soroban-testnet.stellar.org:443 --network-passphrase Test SDF Network ; September 2015
error: unexpected argument 'SDF' found

Usage: stellar network add [OPTIONS] --rpc-url <RPC_URL> --network-passphrase <NETWORK_PASSPHRASE> <NAME>

For more information, try '--help'.
/bin/sh: September: command not found

Here comes the output

Existing Files 
Here comes the output

[ 'target/wasm32-unknown-unknown/release/hello_world.wasm' ]
Deploying Contract hello_world
Here comes the output

JSON file has been saved to ./contracts.json
Error executing command: 
 Command failed: stellar contract bindings typescript --output-dir ./packages/hello_world --network testnet --contract-id CC2L3OKAGKIQZOGP57I76Z7PPVJWTYQ3S6D7SXVQTAMN63TBAFVITWFC
  --wasm target/wasm32-unknown-unknown/release/hello_world.wasm --overwrite
error: --output-dir already exists and you did not specify --overwrite: "./packages/hello_world"
/bin/sh: line 1: --wasm: command not found
```

In your `initialize.js` script, you may want to change [this part](https://github.com/CaptainPrinz/soroban-react-template/blob/2d642ed1aa7e6ad941dead572c1d5a0c43a462eb/initialize.js#L24) to something like this, so the network passphrase gets completely enclosed in quotes.

```js
run(`stellar network add ${process.env.NETWORK} --rpc-url ${process.env.RPC_URL} --network-passphrase "${process.env.NETWORK_PASSPHRASE}"`)
```

There seems to be a trail of other errors in the initialization script that need to be tracked down and figured out. Let me know if you need any help with it.

Author: Fixed

```

This runs the `initialize.js` script and first sets the network profile, account identities, builds the contract `.wasm` file, deploys the contracts, and generates the bindings.

If successful, there should be the following changes in the directory:

1. A new `.soroban` directory, which contains the network profile files, account identity files, and hosted smart contract files in json format
2. A `packages` directory containing the generated TypeScript bindings
3. A `contracts.json` file

You are now finally able to import these generated bindings in your frontend application, create their clients, and use them to interact with the contracts deployed on the network.

> ##Interacting with the contract bindings

Reviewer: ```suggestion
## Interacting with the contract bindings

```

The heading needs a space after `##`, and an empty line after the heading

After running the initialization script, there should be new `packages` directory which confirms successful bindings generation. Navigate to the `packages` directory and there should be a package for the deployed contract(s).
This package will form the last piece in solving the puzzle of interacting with soroban contracts from frontend appications which will be discussed briefly.

To work with the bindings generated, it is recommended the wrap the bindings by creating a new module, instantiating a new client, then finally importing the client and invoking the contract functions.

While it is not necessary to follow the procedure above, you stand to have a better code structure and a much easier way of fixing the code should anything go wrong.

However given the simple structure of our contract, we will proceed to integrate it directly in the frontend without creating a wrapper module.

> ##Setting up the hello world package

Reviewer: ```suggestion
## Setting up the hello world package

```
Same as above

> Navigate to the packages folder to find the hello wold package and observe the directory contents. If the directory does not contain a `dist` folder, you must build the hello world typescript bindings for the package to work in our react app.

Reviewer: ```suggestion
Navigate to the packages folder to find the hello world package and observe the directory contents. If the directory does not contain a `dist` folder, you must build the hello world typescript bindings for the package to work in our react app.
```

```bash
npm run build
```

> After running the `build` command, there should now be a `dist` directory containing `index.js` and `index.ts` files. These will serve as an entry point to the package when it is eventually imported.

Reviewer: I'm not sure the `build` script actually does what you want it to here. In your template repository, `build` is just calling `vite build`, which only concerns itself with the React app, not the contracts themselves.

Author: No In this case I'm referring to the 'build' command in the typescript binding package for the contract. I f you were to observe it's package.json file you'll find the build command

Reviewer: Ah! I see. I must have missed the bit about going to the `packages/hello_world` before this command. Thanks!

> ###Using the hello world package in React

Reviewer: ```suggestion
### Using the hello world package in React

```
Same as above

First we import the following from the package:
1. A `networks` object.
2. A  `Client` class.

```javascript
import { Client, networks } from "../packages/hello_world"
```

Notice how we point to the directory of the package when importing th ebindings by making reference to the packages folder.

Now we instantiate a the `Client` class and pass an object containing the RPC url and a copy of the `networks.testnet` object by using the spread operator.
This technically means that the `Client` class takes in an object of the network configuration along with the contract ID but we make it easier by using the `networks` object.

```javascript
const helloWorldContract = new Client({
    ...networks.testnet,
    rpcUrl: "https://soroban-testnet.stellar.org:443"
  });
```
By printing the `helloWorldContract` object, you'll find the hello method which coincides with the hello function defined in the smart contract. Calling this method will invoke its equivalent in the smart contract.

Finally, to interact with the contract we run the following code:
```javascript
 const { result } = await contract.hello({to: receiver});
```

> ###Modifying the App component

Reviewer: ```suggestion
### Modifying the App component

```
Same as above

> To put it all together, modify the `App` component in the `src` directory by replacing its content with the code below.

Reviewer: ```suggestion
To put it all together, modify the `App.jsx` component in the `src` directory by replacing its content with the code below.
```

```javascript
import { Client, networks } from "../packages/hello_world"
import { useState } from "react";

function App() {
  const contract = new Client({
    ...networks.testnet,
    rpcUrl: "https://soroban-testnet.stellar.org:443"
  });
  console.log(contract);

  const [target, setTarget] = useState("")
  let receiver;
  
  const getResult = async () => {
    const { result } = await contract.hello({to: receiver});
    console.log(result);
    setTarget(result);
  }

  const handleSubmit = (e) => {
    e.preventDefault();
    getResult()
  }

  return (
    <>
      <h1>Response: {target}</h1>
      <form onSubmit={handleSubmit}>
        <input type="text" onChange={(e)=> receiver = e.target.value}/>
        <input type="submit" value="Say hello"/>
      </form>
      
    </>
  )
}

export default App

> 

Reviewer: ```suggestion
```
let's remove the empty line here

```

Now run `npm run dev` to run the project and play with the deployed smart contract from your frontend app. 

## Customization

With the `initialize.js` script being the heart of the template file, you can go through the file and modify the paths in which files are saved if necessary.

Also a great deal of customization is enabled by using environment variables, giving flexibility to change network profiles and account identity without limits.

## Conclusion

> Alot has been covered from creating template files to building, deploying, generating bindings, and interacting with smart contracts. Now go ahead and play witht the code and try applying the procedures to other contracts.

Reviewer: ```suggestion
A lot has been covered from creating template files to building, deploying, generating bindings, and interacting with smart contracts. Now go ahead and play with the code and try applying the procedures to other contracts.
```
Couple spelling nits
